### PR TITLE
fix:  page cache saves Response data before running after filters

### LINKED ISSUE #6281: Bug: secureheaders does not work with cache
### PHP Version

8.1

### CodeIgniter4 Version

4.2.1

### CodeIgniter4 Installation Method

Manual (zip or tar.gz)

### Which operating systems have you tested for this bug?

macOS, Linux

### Which server did you use?

apache

### Database

_No response_

### What happened?

Http headers of "secureheaders" are not present when cache is active

### Steps to Reproduce

1. Active "secureheaders" in "app/Filters.php"

2. Add "$this->cachePage(200);" in your Controller

### Expected Output

The following lines should appear in the http header : 

x-content-type-options: nosniff
x-download-options: noopen
x-frame-options: SAMEORIGIN
x-permitted-cross-domain-policies: none 

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by kenjis**
**Description**
Fixes #6281

The page cache saves Response data before running after filters.
So response headers added by after filters and response body changes by after filters will be lost.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by iRedds** [comment]
Since the issue affects the operation of the SecureHeaders filter, it seems to me that it would also be correct to return the Response in the SecureHeaders::after() method.

**Comment by kenjis** [comment]
What do you mean?
What's the difference between to return the Response in the SecureHeaders::after() and not to return?

**Comment by iRedds** [comment]
@kenjis If we are targeting the implementation of PSR-15 in version 5, then it would be logical to prepare the codebase for the correct behavior. So that it doesn’t work out, as with legacy 3.x, when, with the transfer of the code base from 3.x to 4.x, the code started to work incorrectly or not as intended.

But yes, within the framework of the current concept of the framework, my proposal does not make sense.

**Comment by kenjis** [comment]
@iRedds I got your point. But I think it is related not only SecureHeaders but also all after filters.
The `after()` method should always return `ResponseInterface`.
And if we change the API, at least we must go to 4.3 branch.

We can't do anything in this PR.

**Comment by kenjis** [comment]
Added the upgrade guide.

File: user_guide_src/source/installation/upgrade_422.rst [hunk 1]
```diff
@@ -19,6 +19,17 @@ Mandatory File Changes
 Breaking Changes
 ****************
 
+Web Page Caching Bug Fix
+========================
+
+- :doc:`../general/caching` now caches the Response data after :ref:`after-filters` are executed.
+- For example, if you enables :ref:`secureheaders`, the Response headers are now sent when the page comes from the cache.
```
Comment (MGatner) [inline user_guide_src/source/installation/upgrade_422.rst:26]:
```suggestion
- For example, if you enable :ref:`secureheaders`, the Response headers are now sent when the page comes from the cache.
```

File: user_guide_src/source/installation/upgrade_422.rst [hunk 2]
```diff
@@ -19,6 +19,17 @@ Mandatory File Changes
...
+
+.. important:: If you have written **code based on the bug** that changes to the Response in the after filters are not cached, **sensitive information could be cached and compromised**. If this is the case, change your code to disable caching of the page.
```
Comment (MGatner) [inline user_guide_src/source/installation/upgrade_422.rst:28]:
```suggestion
.. important:: If you have written **code based on this bug** that assumes changes to the Response in "after" filters are not cached then **sensitive information could be cached and compromised**. If this is the case, change your code to disable caching of the page.
```

**Review by MGatner — CHANGES_REQUESTED** [review]
A few typos then looks good! Thanks for adding the explicit docs.

**Review by MGatner — CHANGES_REQUESTED** [review]
A few typos then looks good! Thanks for adding the explicit docs.

**Review by MGatner — APPROVED** [review]

File: system/CodeIgniter.php [hunk 3]
```diff
@@ -298,17 +297,18 @@ protected function initializeKint()
...
-     * @return bool|mixed|RequestInterface|ResponseInterface|void
+     * @return ResponseInterface|void
      */
     public function run(?RouteCollectionInterface $routes = null, bool $returnResponse = false)
     {
         if ($this->context === null) {
             throw new LogicException('Context must be set before run() is called. If you are upgrading from 4.1.x, you need to merge `public/index.php` and `spark` file from `vendor/codeigniter4/framework`.');
         }
 
+        static::$cacheTTL = 0;
```
Comment (samsonasik) [inline system/CodeIgniter.php:310]: is it means that `cacheTTL` value is reset in every run?

Comment (kenjis) [reply system/CodeIgniter.php:310]: Yes. 

$cacheTTL will be set by a Controller.

**Review by samsonasik — APPROVED** [review]

**Pull request merged**